### ISSUE TITLE
Rethink conditional importing with units singleton in units_container.py

### ISSUE THREAD

**Issue by carldlaird** [Pyomo/pyomo#899]
The class PyomoUnitsContainer has functionality that delays the retrieval of the _pint_registry object. This is because the pint package may not be installed, and we do not want the user to get a warning unless they try to use units. This would normally easily be handled with code like the following at the module level:
```python
try:
    import pint as pint_module
except ImportError:
    pint_module = None
```

and then a check in __init__ of the PyomoUnitsContainer class to throw an exception if pint is not installed.

** The problem is: for ease of use on the user, we create a singleton of the PyomoUnitsContainer in the units_container module: see `units = PyomoUnitsContainer()` at the bottom of the module.

Therefore, I instead created a property on the PyomoUnitsContainer to delay checking for the module until it was requested. The logic in this method could be fixed so there was only one check against None each call (put check for pint_module inside the check on `self.__pint_registry is None`).

So... @jsiirola has proposed an alternative that creates a "dummy" PyomoUnitsContainer, called PyomoUnitsContainer_NoPint, and suggests using that for the singleton - this would also delay the error until a user requests use of the units singleton. @jsiirola's suggestion is shown below.

Another alternative that seems a lot cleaner is to push the `units` singleton inside a function call on the module. Then this check would be delayed, and only done once when the user first requested the singleton.

@jsiirola: let's discuss this alternative. Seems easier.

If we decide to go with the suggestion below, then we should make the _NoPint version of the class general (take in the exception message).

Thoughts?

@jsiirola's suggestion.
----------

This has two checks against None every time someone gets the _pint_registry, and could be very expensive (plus the overhead for the property call).  What do you think of the following alternative implementation:
  - Just have the PyomoUnitsContainer directly and unconditionally access the module-level pint_module
  - Swap out the global registry if pint is unavailable.
```python
class PyomoUnitsContainer(object):
    def __init__(self):
        if pint_module is None:
            raise RuntimeError("The PyomoUnitsContainer in the units_container module requires"
                              " the package 'pint', but this package could not be imported."
                              " Please make sure you have 'pint' installed.")
        self._pint_registry = pint_module.UnitRegistry()
    # ....

class PyomoUnitsContainer_NoPint(object):
    def __getattr__(self, attr):
        raise RuntimeError("The PyomoUnitsContainer in the units_container module requires"
                              " the package 'pint', but this package could not be imported."
                              " Please make sure you have 'pint' installed.")
    def __setattr__(self, attr, value):
        raise RuntimeError("The PyomoUnitsContainer in the units_container module requires"
                              " the package 'pint', but this package could not be imported."
                              " Please make sure you have 'pint' installed.")

if pint_module is None:
    units = PyomoUnitsContainer_NoPint()
else:
    units = PyomoUnitsContainer()
```

_Originally posted by @jsiirola in https://github.com/Pyomo/pyomo/pull/833_

**Comment by blnicho** [comment]
@carldlaird any update on this issue? Is it resolved in #1341?

**Comment by carldlaird** [comment]
@blnicho We have not reproduced this issue yet. 
@qtothec Did you get any more information on this? Was there an IDAES import in the file as well?

**Comment by jsiirola** [comment]
Resolved in #1341.
